### PR TITLE
Add GitHub OAuth config

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
[Breaking Change] 
This PR replaces `runWithOidc` parameter with `authType` to support multiple authTypes. 
Defaults to whatever is written in `baseJenkins.yaml`.
With this change we now support github as auth mechanism. 

### Issues Resolved
related #506. Will close this issue once we officially adapt this parameter in internal deployments as well.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: lib/compute/auth-config.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    const githubAuthConfig: { [x: string]: any; } = {
+      github: {
+        githubWebUri: 'https://github.com',
+        githubApiUri: 'https://api.github.com',
+        clientID: 'clientID',
+        clientSecret: 'clientSecret',
+        oauthScopes: 'read:org,user:email',
+      },
+    };
```
Comment (gaiksaya) [inline lib/compute/auth-config.ts:87]: The huge diff is due to file name change. Actual diff only applies to these lines and last 5 files of this file.

File: README.md [hunk 2]
```diff
@@ -144,7 +160,7 @@ You need to restrict access to your jenkins endpoint (load balancer). Here's how
 1. Using command line as below:
 
 ```
-npm run cdk synth OpenSearch-CI-Dev -- -c useSsl=false -c runWithOidc=false -c serverAccessType=ipv4 -c restrictServerAccessTo=10.10.10.10/32
```
Comment (peterzhuamazon) [inline README.md:147]: The runWithOidc is only for federated here? and you dont need authType for synth comment?

File: lib/ci-stack.ts [hunk 3]
```diff
@@ -30,8 +30,8 @@ import { JenkinsWAF } from './security/waf';
 export interface CIStackProps extends StackProps {
   /** Should the Jenkins use https  */
   readonly useSsl?: boolean;
-  /** Should an OIDC provider be installed on Jenkins. */
-  readonly runWithOidc?: boolean;
+  /** Type of login mechanism to adopt */
+  readonly authType?:string;
```
Comment (peterzhuamazon) [inline lib/ci-stack.ts:34]: I see, you just completely change it to a new param.

File: lib/ci-stack.ts [hunk 4]
```diff
@@ -100,18 +100,19 @@ export class CIStack extends Stack {
 
     const useSsl = useSslParameter === 'true';
 
-    const runWithOidcParameter = `${props?.runWithOidc ?? this.node.tryGetContext('runWithOidc')}`;
-    if (runWithOidcParameter !== 'true' && runWithOidcParameter !== 'false') {
-      throw new Error('runWithOidc parameter is required to be set as - true or false');
+    let authType = `${props?.authType ?? this.node.tryGetContext('authType')}`;
+    if (authType.toString() === 'undefined') {
+      authType = 'default';
```
Comment (peterzhuamazon) [inline lib/ci-stack.ts:105]: What is the default type here?

File: lib/compute/auth-config.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        },
+      },
+    };
+
+    jenkinsYaml.jenkins.authorizationStrategy = rolesAndPermissions;
+
+    if (authType === 'github') {
+      jenkinsYaml.jenkins.securityRealm = githubAuthConfig;
+    } else {
+      jenkinsYaml.jenkins.securityRealm = oidcConfig;
```
Comment (peterzhuamazon) [inline lib/compute/auth-config.ts:117]: Is there another oidc setup here apart from github auth with oidc here?

Comment (peterzhuamazon) [reply lib/compute/auth-config.ts:117]: Generic oidc?

File: README.md [hunk 6]
```diff
@@ -126,16 +130,28 @@ $aws secretsmanager put-secret-value \
...
     --secret-id MyTestDatabaseSecret_or_ARN \
     --secret-string file://mycreds.json_or_value
-    ```
-1. Add additional `adminUsers` for role based authentication according to your needs, see [CIStackProps](./lib/ci-stack.ts) for details.
+   ```
+1. Add additional `adminUsers` for role based authentication according to your needs, see [CIStackProps](./lib/ci-stack.ts) for details. 
 1. Run with parameter with one of the following (refer [this](#ssl-configuration) for value of `useSsL`) -
-   1. `npm run cdk deploy OpenSearch-CI-Dev -- -c runWithOidc=true -c useSsl=true` or,
-   1. `cdk deploy OpenSearch-CI-Dev -c runWithOidc=true -c useSsl=true`
+   1. `npm run cdk deploy OpenSearch-CI-Dev -- -c authType=oidc/github/default -c useSsl=true` or,
```
Comment (rishabh6788) [inline README.md:153]: nit: `oidc|github|default`

Comment (rishabh6788) [reply README.md:153]: Jenkins github is OIDC plugin right?

File: lib/ci-stack.ts (hunk 4 above)
Comment (rishabh6788) [reply lib/ci-stack.ts:105]: I believe it is the plain old user/pass mechanism that jenkins provides.

File: README.md (hunk 2 above)
Comment (gaiksaya) [reply README.md:147]: Nope. removed the runWithOidc instances in readme

File: lib/ci-stack.ts (hunk 3 above)
Comment (gaiksaya) [reply lib/ci-stack.ts:34]: Yes!

File: lib/ci-stack.ts (hunk 4 above)
Comment (gaiksaya) [reply lib/ci-stack.ts:105]: Whatever is the base_jenkins.yaml. @rishabh6788 is right by default is user/pass as before

File: lib/compute/auth-config.ts (hunk 5 above)
Comment (gaiksaya) [reply lib/compute/auth-config.ts:117]: Yes! Generic one where users can use google, msft, etc.

**Review by rishabh6788 — APPROVED** [review]

**Pull request merged**